### PR TITLE
add benchmark imagenet code example

### PR DESCRIPTION
**Pull request by andrewkho**
Adds Imagenet benchmarking to examples/nodes

Test plan, tried a handful of options with Python 3.13t build on Linux eg:
```
python -Xgil=0 examples/nodes/imagenet_benchmark.py -d ~/local/imagenet_1k/train/ --max-steps 1000 --loader thread --shuffle=True --in-order=True
Running benchmark with args=Namespace(loader='thread', num_workers=4, batch_size=16, in_order=True, shuffle=True, max_steps=1000, max_duration=60, warmup_steps=0, pin_memory=False, imagenet_dir='/home/andrewkh/local/imagenet_1k/train/')...
create iter took 0.020316815935075283 seconds
0 / 1000, 5.683861672878265e-06 seconds elapsed
100 / 1000, 1.8675199393182993 seconds elapsed
200 / 1000, 3.7960099279880524 seconds elapsed
300 / 1000, 5.413591359741986 seconds elapsed
400 / 1000, 7.112115032039583 seconds elapsed
500 / 1000, 8.72194852028042 seconds elapsed
600 / 1000, 10.356252245604992 seconds elapsed
700 / 1000, 12.0448250900954 seconds elapsed
800 / 1000, 13.797063117846847 seconds elapsed
900 / 1000, 15.468625104986131 seconds elapsed
================================================================================
args=Namespace(loader='thread', num_workers=4, batch_size=16, in_order=True, shuffle=True, max_steps=1000, max_duration=60, warmup_steps=0, pin_memory=False, imagenet_dir='/home/andrewkh/local/imagenet_1k/train/')
Benchmark complete, 999 steps took 17.021807146258652 seconds, for a total of 15984 images
939.0307305598419 img/sec, 58.68942065999012 batches/sec
create_iter_dt=0.020316815935075283, warmup_dt=1.7201527953147888e-06, sum((create_iter_dt, warmup_dt, iter_time))=17.042125682346523
```

Also tried:
```
 python examples/nodes/imagenet_benchmark.py -d ~/local/imagenet_1k/train/ --max-steps 1000 --loader classic
python examples/nodes/imagenet_benchmark.py -d ~/local/imagenet_1k/train/ --max-steps 1000 --loader process --shuffle=True
 python examples/nodes/imagenet_benchmark.py -d ~/local/imagenet_1k/train/ --max-steps 1000 --loader process --shuffle=True --in-order=True
```

### CODE REVIEW

**Comment by andrewkho** [comment]
cc @SonicField

**Review by ramanishsingh — APPROVED** [review]

File: examples/nodes/imagenet_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        help="Whether to use multi-threaded parallelism, multi-process parallelism, or the classic torch.utils.data.DataLoader (multi-process only)",
+    )
+    parser.add_argument(
+        "--num-workers",
+        type=int,
+        default=4,
+        help="Number of workers to parallelize with",
+    )
+    parser.add_argument("--batch-size", type=int, default=16, help="Batch size for dataloading")
+    parser.add_argument("--in-order", type=bool, default=False, help="Whether to enforce ordering")
```
Comment (divyanshk) [inline examples/nodes/imagenet_benchmark.py:179]: Should the default be `true` ?

File: examples/nodes/imagenet_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    def __len__(self):
+        return len(self.imagenet_data)
+
+    def __getitem__(self, i: int) -> dict:
+        return self.tx(self.imagenet_data[i])
+
+
+def setup_classic(args):
+    dataset = ImagenetDataset(args.imagenet_dir)
+    assert args.in_order is False, "torch.utils.data.DataLoader does not support out-of-order iteration yet!"
```
Comment (divyanshk) [inline examples/nodes/imagenet_benchmark.py:84]: should the assert be args.in_order is true ? also, does the string have to change with https://github.com/pytorch/pytorch/pull/141833 ?

**Comment by divyanshk** [comment]
In case folks are interested in the results, should we paste the results in here? Thanks.

File: examples/nodes/imagenet_benchmark.py (hunk 1 above)
Comment (andrewkho) [reply examples/nodes/imagenet_benchmark.py:179]: Fastest results are when False

File: examples/nodes/imagenet_benchmark.py (hunk 2 above)
Comment (andrewkho) [reply examples/nodes/imagenet_benchmark.py:84]: We can drop this after the next release of pytorch, and/or do a version check

**Pull request merged**